pkg/oidc/token.go

```diff
@@ -3,331 +3,433 @@ package oidc
-	locale, _ := language.Parse(j.Locale)
-	return UserinfoProfile{
-		Name:              j.Name,
-		GivenName:         j.GivenName,
-		FamilyName:        j.FamilyName,
-		MiddleName:        j.MiddleName,
-		Nickname:          j.Nickname,
-		Profile:           j.Profile,
-		Picture:           j.Picture,
-		Website:           j.Website,
-		Gender:            Gender(j.Gender),
-		Birthdate:         j.Birthdate,
-		Zoneinfo:          j.Zoneinfo,
-		Locale:            locale,
-		UpdatedAt:         time.Unix(j.UpdatedAt, 0).UTC(),
-		PreferredUsername: j.PreferredUsername,
-	}
+type AccessTokenResponse struct {
+	AccessToken  string `json:"access_token,omitempty" schema:"access_token,omitempty"`
+	TokenType    string `json:"token_type,omitempty" schema:"token_type,omitempty"`
+	RefreshToken string `json:"refresh_token,omitempty" schema:"refresh_token,omitempty"`
+	ExpiresIn    uint64 `json:"expires_in,omitempty" schema:"expires_in,omitempty"`
+	IDToken      string `json:"id_token,omitempty" schema:"id_token,omitempty"`
 }
 
-func (j *jsonToken) UnmarshalUserinfoEmail() UserinfoEmail {
-	return UserinfoEmail{
-		Email:         j.Email,
-		EmailVerified: j.EmailVerified,
-	}
+type JWTProfileAssertion struct {
+	PrivateKeyID string   `json:"-"`
+	PrivateKey   []byte   `json:"-"`
+	Issuer       string   `json:"issuer"`
+	Subject      string   `json:"sub"`
+	Audience     Audience `json:"aud"`
+	Expiration   Time     `json:"exp"`
+	IssuedAt     Time     `json:"iat"`
 }
 
-func (j *jsonToken) UnmarshalUserinfoPhone() UserinfoPhone {
-	return UserinfoPhone{
-		PhoneNumber:         j.Phone,
-		PhoneNumberVerified: j.PhoneVerified,
+func NewJWTProfileAssertionFromKeyJSON(filename string, audience []string) (*JWTProfileAssertion, error) {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
 	}
+	return NewJWTProfileAssertionFromFileData(data, audience)
 }
 
-func (j *jsonToken) UnmarshalUserinfoAddress() *UserinfoAddress {
-	if j.JsonUserinfoAddress == nil {
-		return nil
+func NewJWTProfileAssertionFromFileData(data []byte, audience []string) (*JWTProfileAssertion, error) {
+	keyData := new(struct {
+		KeyID  string `json:"keyId"`
+		Key    string `json:"key"`
+		UserID string `json:"userId"`
+	})
+	err := json.Unmarshal(data, keyData)
+	if err != nil {
+		return nil, err
 	}
-	return &UserinfoAddress{
-		Country:       j.JsonUserinfoAddress.Country,
-		Formatted:     j.JsonUserinfoAddress.Formatted,
-		Locality:      j.JsonUserinfoAddress.Locality,
-		PostalCode:    j.JsonUserinfoAddress.PostalCode,
-		Region:        j.JsonUserinfoAddress.Region,
-		StreetAddress: j.JsonUserinfoAddress.StreetAddress,
+	return NewJWTProfileAssertion(keyData.UserID, keyData.KeyID, audience, []byte(keyData.Key)), nil
+}
+
+func NewJWTProfileAssertion(userID, keyID string, audience []string, key []byte) *JWTProfileAssertion {
+	return &JWTProfileAssertion{
+		PrivateKey:   key,
+		PrivateKeyID: keyID,
+		Issuer:       userID,
```

Comment:
userID in issuer?

Reply:
jwt profile is done by service accounts... so they're actually the issuer